src/test/ensemble_flow_custom_node_tests.cpp

```diff
@@ -0,0 +1,974 @@
+            "outputs": [
+                {"pipeline_output": {"node_name": "custom_node",
+                                     "data_item": "custom_node_output"}
+                }
+            ]
+        }
+    ]
+})";
+
+TEST_F(EnsembleFlowCustomNodeLoadConfigThenExecuteTest, ReferenceMissingLibraryThenCorrect) {
+    std::unique_ptr<Pipeline> pipeline;
+    this->prepareRequest(inputValues);
+
+    this->loadCorrectConfiguration();
+    ASSERT_EQ(manager.createPipeline(pipeline, pipelineName, &request, &response), StatusCode::OK);
+    ASSERT_EQ(pipeline->execute(), StatusCode::OK);
+    this->checkResponseForCorrectConfiguration();
+    response.Clear();
+
+    this->loadConfiguration(pipelineCustomNodeReferenceMissingLibraryConfig);
+    ASSERT_EQ(manager.createPipeline(pipeline, pipelineName, &request, &response), StatusCode::PIPELINE_DEFINITION_NOT_LOADED_YET);
+    response.Clear();
+
+    this->loadCorrectConfiguration();
+    ASSERT_EQ(manager.createPipeline(pipeline, pipelineName, &request, &response), StatusCode::OK);
+    ASSERT_EQ(pipeline->execute(), StatusCode::OK);
+    this->checkResponseForCorrectConfiguration();
+}
+
+static const char* pipelineCustomNodeReferenceLibraryWithExecutionErrorLibraryConfig = R"(
+{
+    "model_config_list": [],
+    "custom_node_library_config_list": [
+        {
+            "name": "lib_add_sub_new",
+            "base_path": "/ovms/bazel-bin/src/lib_node_mock.so"
+        }
+    ],
+    "pipeline_config_list": [
+        {
+            "name": "my_pipeline",
+            "inputs": ["pipeline_input"],
+            "nodes": [
+                {
+                    "name": "custom_node",
+                    "library_name": "lib_add_sub_new",
+                    "params": {
+                        "add_value": "3.2",
+                        "sub_value": "2.7"
+                    },
+                    "type": "custom",
+                    "inputs": [
+                        {"input_numbers": {"node_name": "request",
+                                           "data_item": "pipeline_input"}}
+                    ],
+                    "outputs": [
+                        {"data_item": "output_numbers",
+                         "alias": "custom_node_output"}
+                    ]
+                }
+            ],
+            "outputs": [
+                {"pipeline_output": {"node_name": "custom_node",
+                                     "data_item": "custom_node_output"}
+                }
+            ]
+        }
+    ]
+})";
+
+TEST_F(EnsembleFlowCustomNodeLoadConfigThenExecuteTest, ReferenceLibraryWithExecutionErrorThenCorrect) {
+    std::unique_ptr<Pipeline> pipeline;
+    this->prepareRequest(inputValues);
+
+    this->loadCorrectConfiguration();
+    ASSERT_EQ(manager.createPipeline(pipeline, pipelineName, &request, &response), StatusCode::OK);
+    ASSERT_EQ(pipeline->execute(), StatusCode::OK);
+    this->checkResponseForCorrectConfiguration();
+    response.Clear();
```

Comment:
Is this necessary? I think this library just return wrong ret codes? 

I think that you anticipated future config validation that could potentialy make this test invalid without those lines. So maybe add comment why those are necessary?

Reply:
Every step (divided by empty line) is treated as separate gRPC/REST API request. To make sure we have empty PredictResponse (as we do on real OVMS path for every request) we clear it so we can reuse the same variable.

Reply:
Will add comment describing why we need to load correctly first.